### DISCUSSION TITLE
text segments containing "%"

### DISCUSSION THREAD

**Discussion by amoldeshpande** [JanDeDobbeleer/oh-my-posh#937] [Troubleshoot]
I have text segments like this
```
		{
		  "type": "text",
		  "style": "powerline",
		  "foreground": "#ffffff",
		  "background": "#0077c2",
		  "properties": {
			"text" : "%c03"
		  }
		},

```
With the current version of posh, the `%c03` gets turned into `%%c03%`, which causes `tcsh` to treat the `%` as escaped and disregard the prompt specification.

What is the workaround for the case where I need the `%` literally in the text segment ?

thanks.

**Comment by JanDeDobbeleer** [comment]
To be clear, you want tcsh to replace it? Because that we don't support.

**Comment by amoldeshpande** [comment]
Sorry, I don't understand your question.   I need posh to generate an unescaped `%c03`  in the text segment.  It used to work before and doesn't any more.

**Reply by JanDeDobbeleer** [reply]
Thx. @lnu you did adjust this a little while back. Any feedback?

**Reply by lnu** [reply]
I did a fix to avoid variable expansion [here](https://github.com/JanDeDobbeleer/oh-my-posh/commit/3f02ad4eb30fb91535d6c105f7b8f9ed0865f14b) for this [reason](https://github.com/JanDeDobbeleer/oh-my-posh/pull/892)  
What do we want to do with variable expansion: support it everywhere(postfix, prefix,...) or only in the text segment? Because without the fix any % character with an existing variable will be expanded and sometimes where you don't expect it :).  
Moreover, since % is only applicable to zsh, you cannot really have a cross shell theme or you will end with variable names in pwsh and expanded variables in zsh.

**Reply by JanDeDobbeleer** [reply]
I would argue that for a literal %, a glyph can be used instead. Otherwise we enter a world of pain.

**Reply by lnu** [reply]
I don't know what's best... but env variables are available through templating:  
```
 "text": "{{.Folder}}"
```

So maybe it's not as customizable as using prompt variables... but if you're using omp it's to have the same shell everywhere and to help you built it, without relying on specific variables/sequences from shells?

**Reply by amoldeshpande** [reply]
ok, based on the suggestion I was able to use `\u0025` instead of `%` .  I would rather have a way to have literal text segments (perhaps a segment called `literal_text` )  than this hack, but I guess it works for me. thanks.

**Reply by lnu** [reply]
I'm not sure to have understood exactly what you wanted to do... could you put a screenshot of your actual prompt and a copy of your theme?

**Comment by amoldeshpande** [comment]
The relevant segment of the theme is up in the first message.  I can't seem to upload an image, but here's the summary:

Ignoring posh for a second, if I set my `tcsh` prompt:
```
set ellipsis
set prompt='%c03 >'
```
It results in  the following:
```
J:/github/UnrealEngine/Engine >cd Source/
J:...UnrealEngine/Engine/Source >echo "$prompt"
%c03 >
 J:...UnrealEngine/Engine/Source >
```
Basically, tcsh is keeping the last 3 components of the path and replacing the preceding with an ellipsis.

So, I was trying to achieve the same thing with posh by adding a `text` segment to the theme, as I indicated in my first comment.  This used to give me a  literal `%c03` in the string for the prompt spec.

I need the ellipsis behavior as it keeps the prompt to a manageable size.  I don't think `.Folder` offers the same functionality.   Even if it does, the core issue is that there are many other `%`-using prompt specifications that `tcsh` (and probably `zsh`) uses, and having to specify those with a unicode sequence feels hacky to me.  

Hope that makes sense.

**Reply by lnu** [reply]
I can see two issues here. You want to use prompt variable expansion to solve the second one: displaying the current directory in a specific format.  
Using specific shell prompt variables is not possible right now, since `%` is escaped. Maybe it's something that can be changed.  
For the second one, you can use the `path` segment to display the current path(different formats available).  
Thanks for the clarification.

**Reply by JanDeDobbeleer** [reply]
@lnu I want to keep that expansion unsupported as that opens a can of worms. If there's a need to have a similar behaviour, we have to build it into the relevant segment (folder in this case).

**Reply by lnu** [reply]
"Can of worms" => exactly :smiley:

**Reply by JanDeDobbeleer** [reply]
@amoldeshpande can you open a feature request to add the `%c03` style to the path segment?

**Reply by amoldeshpande** [reply]
I would much rather have a feature to have a literal text segment or a flag to the text segment to turn off escaping.

**Reply by amoldeshpande** [reply]
BTW I was wrong about being able to use \u0025 instead of `%`. I must have gotten confused about which version of posh I was using. too many permutations :(

**Reply by lnu** [reply]
What’s the issue with the \u0025?

**Reply by amoldeshpande** [reply]
it just resolves to `%`  (which makes sense) so is also escaped.

**Reply by JanDeDobbeleer** [reply]
Ah yes. It's a literal percentage. Because the above reasons we will not be adding support for this. I will however extend the path styles to also display paths like this.

**Reply by amoldeshpande** [reply]
can you explain why you are opposed to a literal passthrough option with a flag ?  

 If you check the `tcsh` man page at https://linux.die.net/man/1/tcsh and search for the prompt variable section (or search for `%`  and you'll find the section) there are  several different  prompt expansions settable with `%`.   

`zsh` https://zsh.sourceforge.io/Doc/Release/Prompt-Expansion.html#Prompt-Expansion  has even more. 

It seems self-defeating to emulate a subset with other mechanisms when the shells themselves are perfectly capable of  handling these things.  

You could add a flag to the text segment like `no_escapes:true` or whatever .  This would turn off interpretation of the text segment and pass it as-is to the shell. If a user wants to do that, why not support it ?

**Reply by JanDeDobbeleer** [reply]
Because we escape when the full prompt is rendered to avoid having to deal with knowing intent or not (for example when used in templates). That context isn't available to the renderer (or the segment tbh). I can explore the path to use a replacement character for this, and then substitute that after we escape for unwanted side effects, but also there issues can happen. I'll give that strategy a go.
